### PR TITLE
Abstract out the command line options for block assembly

### PR DESCRIPTION
**Pull request by sipa**
It feels strange that `BlockAssembler`'s constructor inspects command-line options to set its configurable fields. This PR introduces a BlockAssembler::Options struct which can be constructed from the command-line options, but is constructed with static values for tests instead.

### CODE REVIEW

File: src/miner.cpp [hunk 1]
```diff
@@ -72,43 +72,56 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
...
+    nBlockMaxSize = DEFAULT_BLOCK_MAX_SIZE;
+}
+
+BlockAssembler::BlockAssembler(const CChainParams& params, const Options& options) : chainparams(params)
+{
+    blockMinFeeRate = options.blockMinFeeRate;
+    // Limit weight to between 4K and MAX_BLOCK_WEIGHT-4K for sanity:
+    nBlockMaxWeight = std::max<size_t>(4000, std::min<size_t>(MAX_BLOCK_WEIGHT - 4000, options.nBlockMaxWeight));
+    // Limit size to between 1K and MAX_BLOCK_SERIALIZED_SIZE-1K for sanity:
+    nBlockMaxSize = std::max<size_t>(1000, std::min<size_t>(MAX_BLOCK_SERIALIZED_SIZE - 1000, options.nBlockMaxSize));
```
Comment (luke-jr) [inline src/miner.cpp]: Why not do the `std::max` on the `Options` object once, and just have a reference to that object on the `BlockAssembler`?

File: src/miner.cpp [hunk 2]
```diff
@@ -72,43 +72,56 @@ int64_t UpdateTime(CBlockHeader* pblock, const Consensus::Params& consensusParam
     return nNewTime - nOldTime;
 }
 
-BlockAssembler::BlockAssembler(const CChainParams& _chainparams)
-    : chainparams(_chainparams)
+BlockAssembler::Options::Options() {
+    blockMinFeeRate = CFeeRate(DEFAULT_BLOCK_MIN_TX_FEE);
+    nBlockMaxWeight = DEFAULT_BLOCK_MAX_WEIGHT;
+    nBlockMaxSize = DEFAULT_BLOCK_MAX_SIZE;
```
Comment (luke-jr) [inline src/miner.cpp]: Is there ever a time we would want to just assume the hard-coded defaults? Leaving the fields uninitialised has the benefit of exposing bugs in static analysis and/or valgrind/memorysanitizer.

**Review by jonasschnelli — APPROVED** [review]
utACK 277b472fb21db824fea9625a0287640fab4df282

**Comment by sdaftuar** [comment]
ACK 277b472

**Comment by laanwj** [comment]
> It feels strange that BlockAssembler's constructor inspects command-line options to set its configurable fields. 

Yes that is strange, utACK 277b472

**Comment by theuni** [comment]
post-merge utACK. I love these kinds of separations!

**Pull request merged**